common/map.h

```diff
@@ -0,0 +1,462 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_COMMON_MAP_H_
+#define CARBON_COMMON_MAP_H_
+
+#include <algorithm>
+#include <concepts>
+#include <type_traits>
+#include <utility>
+
+#include "common/check.h"
+#include "common/raw_hashtable.h"
+#include "llvm/Support/Compiler.h"
+
+namespace Carbon {
+
+// Forward declarations to resolve cyclic references.
+template <typename KeyT, typename ValueT>
+class MapView;
+template <typename KeyT, typename ValueT>
+class MapBase;
+template <typename KeyT, typename ValueT, ssize_t MinSmallSize>
+class Map;
+
+// A read-only view type for a map from key to value.
+//
+// This view is a cheap-to-copy type that should be passed by value, but
+// provides view or read-only reference semantics to the underlying map data
+// structure.
+//
+// This should always be preferred to a `const`-ref parameter for the `MapBase`
+// or `Map` type as it provides more flexibility and a cleaner API.
+//
+// Note that while this type is a read-only view, that applies to the underlying
+// *map* data structure, not the individual entries stored within it. Those can
+// be mutated freely. If we applied a deep-`const` design here, it would prevent
```

Comment:
And equality. Added to the comment.

I'm slightly tempted to make the keys `const` throughout as I view mutating those as the least plausible use case I can imagine... And restricting it has some value in terms of tightening the API.

Alternatively, I could imagine shifting the API to return keys *by value* from accessors and constructively preclude mutation as a consequence. That would have even more value in terms of tightening the API as that would allow us to rotate the underlying key storage when desirable.

Open to thoughts about this.

Reply:
That seems tempting, but inconsistent with the situation with sets, where I do expect that some users will want to mutate the "key", since it will be the whole entry.